File: packages/types/src/tool/execution.ts
/**
 * The tool implements the relevant type definitions
 */

import type { ID, Timestamp } from "../common.js";

/**
 * Tool Call Record Type
 */
export interface ToolCall {
  /** Tool Call ID */
  id: ID;
  /** Tool ID */
  toolId: ID;
  /** Tool name */
  toolName?: string;
  /** Tool parameters */
  parameters: Record<string, unknown>;
  /** invocation result */
  result?: unknown;
  /** error message */
  error?: unknown;
  /** invocation time */
  timestamp: Timestamp;
  /** Execution time (milliseconds) */
  executionTime?: Timestamp;
}

/**
 * Tool implementation options
 */
export interface ToolExecutionOptions {
  /** Timeout time (milliseconds) */
  timeout?: number;
  /** Maximum number of retries */
  retries?: number;
  /** Retry delay (milliseconds) */
  retryDelay?: number;
  /** Whether to enable exponential backoff */
  exponentialBackoff?: boolean;
  /** Abort signal (for canceling execution) */
  signal?: AbortSignal;
}

/**
 * Tool execution results (SDK internal use)
 * Contains complete execution metadata for internal streaming
 */
export interface ToolExecutionResult {
  /** success or failure */
  success: boolean;
  /** Implementation results (raw data) */
  result?: unknown;
  /** error message */
  error?: string;
  /** Execution time (milliseconds) */
  executionTime: number;
  /** Retries */
  retryCount: number;
}

/**
 * Tool output (used by the tool implementation layer)
 * Tool execute function return value type
 */
export interface ToolOutput {
  /** success or failure */
  success: boolean;
  /** Output content (can be of any type, will eventually be serialized as a string and sent to LLM) */
  content: unknown;
  /** error message */
  error?: string;
}

File: crates/foundation/wf-types/src/tool/execution.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolCall {
    pub id: String,
    pub tool_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_name: Option<String>,
    pub parameters: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub timestamp: super::super::Timestamp,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub execution_time: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolExecutionOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub retries: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub retry_delay: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exponential_backoff: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolExecutionResult {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    pub execution_time: i64,
    pub retry_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolOutput {
    pub success: bool,
    pub content: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<crate::Metadata>,
}
